File: Tree.java
import java.util.LinkedList;
import java.util.Queue;

public class Tree {

    public static class Node {
        int data;
        Node left, right;

        Node(int data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }
    }

    public static class BinaryTree {
        private static int index = -1;

        public static Node buildTree(int[] nodes) {
            index++;
            if (index >= nodes.length || nodes[index] == -1) {
                return null;
            }

            Node newNode = new Node(nodes[index]);
            newNode.left = buildTree(nodes);
            newNode.right = buildTree(nodes);

            return newNode;
        }

        public static void preOrder(Node root) {
            if (root == null) {
                return;
            }
            System.out.print(root.data + " ");
            preOrder(root.left);
            preOrder(root.right);
        }

        public static void inOrder(Node root) {
            if (root == null) {
                return;
            }
            inOrder(root.left);
            System.out.print(root.data + " ");
            inOrder(root.right);
        }

        public static void postOrder(Node root) {
            if (root == null) {
                return;
            }
            postOrder(root.left);
            postOrder(root.right);
            System.out.print(root.data + " ");
        }

        public static int countNodes(Node root) {
            if (root == null) {
                return 0;
            }
            int leftNode = countNodes(root.left);
            int rightNode = countNodes(root.right);
            return leftNode + rightNode + 1;
        }

        public static int sumNodes(Node root) {
            if (root == null) {
                return 0;
            }
            int leftSum = sumNodes(root.left);
            int rightSum = sumNodes(root.right);
            return leftSum + rightSum + root.data;
        }

        static class TreeInfo {
            int ht;
            int dia;

            TreeInfo(int ht, int dia) {
                this.ht = ht;
                this.dia = dia;
            }
        }

        public static int height(Node root) {
            if (root == null) {
                return 0;
            }
            int leftHeight = height(root.left);
            int rightHeight = height(root.right);
            int myHeight = Math.max(leftHeight, rightHeight) + 1;
            return myHeight;
        }

        public static int diameter(Node root) {
            if (root == null) {
                return 0;
            }

            int dia1 = diameter(root.left);
            int dia2 = diameter(root.right);
            int dia3 = height(root.left) + height(root.right) + 1;

            return Math.max(Math.max(dia1, dia2), dia3);
        }

        public static TreeInfo diameter2(Node root) {
            if (root == null) {
                return new TreeInfo(0, 0);
            }

            TreeInfo left = diameter2(root.left);
            TreeInfo right = diameter2(root.right);

            int myHeight = Math.max(left.ht, right.ht) + 1;

            int dia1 = left.dia;
            int dia2 = right.dia;
            int dia3 = left.ht + right.ht + 1;

            int myDiameter = Math.max(Math.max(dia1, dia2), dia3);

            return new TreeInfo(myHeight, myDiameter);
        }

        public static void levelOrder(Node root) {
            Queue<Node> q = new LinkedList<>();
            q.add(root);
            q.add(null);
            if (root == null) {
                return;
            }
            while (!q.isEmpty()) {
                Node currNode = q.remove();
                if (currNode == null) {
                    System.out.println();
                    if (q.isEmpty()) {
                        break;
                    } else {
                        q.add(null);
                    }
                } else {
                    System.out.print(currNode.data + " ");
                    if (currNode.left != null) {
                        q.add(currNode.left);
                    }
                    if (currNode.right != null) {
                        q.add(currNode.right);
                    }
                }
            }
        }

        public static int kSum(Node root,int k) {
            if (root == null) {
                return 0;
            }
            Queue<Node> q = new LinkedList<>();
            int sum = 0; 
            int level=0;
            q.add(root);
            
            while (!q.isEmpty()) {
                int size=q.size();
                for (int i = 0; i < size; i++) {
                    Node currNode = q.remove();
                    if(level==k){
                        sum+=currNode.data;
                    }
                    if(currNode.left!=null){
                        q.add(currNode.left);
                    }
                    if(currNode.right!=null){
                        q.add(currNode.right);
                    }
                } 
                if(level==k){
                    break;
                    
                }
                level++;
            }
            
            return sum;
        }
    }

    public static void main(String[] args) {
        int nodes[] = { 1, 2, 4, -1, -1, 5, -1, -1, 3, -1, 6, -1, -1 };
        Node root = BinaryTree.buildTree(nodes);
        System.out.println("Root node: " + root.data);

        System.out.print("Pre-order traversal: ");
        BinaryTree.preOrder(root);

        System.out.println();
        System.out.print("In-order traversal: ");
        BinaryTree.inOrder(root);
        System.out.println();
        System.out.print("Postorder traversal: ");
        BinaryTree.postOrder(root);

        System.out.println();
        System.out.println("Level order traversal: ");
        BinaryTree.levelOrder(root);

        System.out.println("Number of Nodes: " + BinaryTree.countNodes(root));
        System.out.println("Sum of nodes = " + BinaryTree.sumNodes(root));

        int height = BinaryTree.height(root);
        int diameter = BinaryTree.diameter(root);

        System.out.println("Height of tree: " + height);
        System.out.println("Diameter of tree: " + diameter);

        BinaryTree.TreeInfo treeInfo = BinaryTree.diameter2(root);
        System.out.println("Diameter of tree (from diameter2 method): " + treeInfo.dia);
        //int k=5;
        System.out.println(BinaryTree.kSum(root,1));
    }
}

File: Tree.cpp
#include <iostream>
#include <cmath>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;

    Node(int data) {
        this->data = data;
        left = nullptr;
        right = nullptr;
    }
};

// Function declarations
Node* insert(Node* root, int data);
void inorder(Node* root);
bool search(Node* root, int data);
int height(Node* root);
void levelOrder(Node* root);
void printLevel(Node* root, int level);
Node* deleteNode(Node* root, int data);
Node* minValueNode(Node* node);
Node* mirror(Node* root);
void printTree(Node* root);
void printLevelStructure(Node* root, int level, int space);

Node* insert(Node* root, int data) {
    if (root == nullptr) {
        return new Node(data);
    }
    if (data < root->data) {
        root->left = insert(root->left, data);
    } else {
        root->right = insert(root->right, data);
    }
    return root;
}

void inorder(Node* root) {
    if (root == nullptr) return;
    inorder(root->left);
    cout << root->data << " ";
    inorder(root->right);
}

bool search(Node* root, int data) {
    if (root == nullptr) return false;
    if (root->data == data) return true;
    if (data < root->data) return search(root->left, data);
    return search(root->right, data);
}

int height(Node* root) {
    if (root == nullptr) return 0;
    int leftHeight = height(root->left);
    int rightHeight = height(root->right);
    return 1 + max(leftHeight, rightHeight);
}

void levelOrder(Node* root) {
    int h = height(root);
    for (int i = 1; i <= h; i++) {
        printLevel(root, i);
        cout << endl;
    }
}

void printLevel(Node* root, int level) {
    if (root == nullptr) return;
    if (level == 1) {
        cout << root->data << " ";
    } else if (level > 1) {
        printLevel(root->left, level - 1);
        printLevel(root->right, level - 1);
    }
}

Node* deleteNode(Node* root, int data) {
    if (root == nullptr) return root;

    if (data < root->data) {
        root->left = deleteNode(root->left, data);
    } else if (data > root->data) {
        root->right = deleteNode(root->right, data);
    } else {
        // Node to be deleted found
        if (root->left == nullptr) {
            Node* temp = root->right;
            delete root;
            return temp;
        } else if (root->right == nullptr) {
            Node* temp = root->left;
            delete root;
            return temp;
        }

        // Node with two children: get the inorder successor (smallest in the right subtree)
        Node* temp = minValueNode(root->right);

        // Copy the inorder successor's content to this node
        root->data = temp->data;

        // Delete the inorder successor
        root->right = deleteNode(root->right, temp->data);
    }
    return root;
}

Node* minValueNode(Node* node) {
    Node* current = node;
    while (current && current->left != nullptr) {
        current = current->left;
    }
    return current;
}

Node* mirror(Node* root) {
    if (root == nullptr) return root;
    swap(root->left, root->right);
    mirror(root->left);
    mirror(root->right);
    return root;
}

void printTree(Node* root) {
    if (root == nullptr) {
        cout << "Tree is empty." << endl;
        return;
    }

    int treeHeight = height(root);
    int space = pow(2, treeHeight) - 1;

    for (int level = 1; level <= treeHeight; level++) {
        printLevelStructure(root, level, space);
        cout << endl;
        space /= 2;
    }
}

void printLevelStructure(Node* root, int level, int space) {
    if (root == nullptr) {
        for (int i = 0; i < space * 2; i++) {
            cout << " "; 
        }
        return;
    }

    if (level == 1) {
        for (int i = 0; i < space; i++) {
            cout << " "; 
        }
        cout << root->data;
        for (int i = 0; i < space; i++) {
            cout << " "; 
        }
    } else {
        printLevelStructure(root->left, level - 1, space);
        printLevelStructure(root->right, level - 1, space);
    }
}

int main() {
    Node* root = nullptr;

    root = insert(root, 50);
    root = insert(root, 30);
    root = insert(root, 20);
    root = insert(root, 40);
    root = insert(root, 70);
    root = insert(root, 60);
   
    cout << "Tree Structure:" << endl;
    printTree(root);

    cout << "\nIn-order traversal: ";
    inorder(root);
    cout << endl;

    int searchValue = 40;
    if (search(root, searchValue)) {
        cout << "Node " << searchValue << " found!" << endl;
    } else {
        cout << "Node " << searchValue << " not found!" << endl;
    }

    cout << "\nHeight of the tree: " << height(root) << endl;

    cout << "\nLevel-order traversal:" << endl;
    levelOrder(root);

    int deleteValue = 20;
    cout << "\nDeleting node " << deleteValue << "..." << endl;
    root = deleteNode(root, deleteValue);

    cout << "\nTree Structure after deletion:" << endl;
    printTree(root);

    root = mirror(root);
    cout << "\nMirrored Tree Structure:" << endl;
    printTree(root);

    return 0;
}
